src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/CSharpVirtualDocument.cs

```diff
@@ -82,6 +90,42 @@ public override VirtualDocumentSnapshot Update(IReadOnlyList<TextChange> changes
             return _currentSnapshot;
         }
 
-        private CSharpVirtualDocumentSnapshot UpdateSnapshot() => new CSharpVirtualDocumentSnapshot(Uri, TextBuffer.CurrentSnapshot, HostDocumentSyncVersion);
+        private bool TryRevertProvisionalChanges()
+        {
+            if (!_hasProvisionalChanges)
+            {
+                return false;
+            }
+
+            Debug.Assert(_previousSnapshot != null);
+
+            using var revertEdit = TextBuffer.CreateEdit(EditOptions.None, _previousSnapshot.Snapshot.Version.VersionNumber, InviolableEditTag.Instance);
```

Comment:
lol wth is an `InviolableEditTag`s purpose

Reply:
I added a comment on the class. It is basically to prevent someone else listening to this TextBuffer's changes from making changes in response to this edit.